### ISSUE TITLE
java.lang.IllegalStateException: Type names equal for class LongFieldType and TTLFieldType

### ISSUE THREAD

**Issue by rileytj** [elastic/elasticsearch#14566]
In class MappedFieldType method checkTypeName the above exception is thrown in the cases when the other class is an inherited child of the parent class as in the case of the TTLFieldType which inherits LongFieldType.

This check

 } else if (getClass() != other.getClass()) {

should be changed to maybe "instance of"

https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

**Comment by rjernst** [comment]
This needs to stay an exact equality check. ttl and long are not exactly the same. They could have different underlying field type settings available. The data types of fields must be exactly identical across document types.

**Comment by rileytj** [comment]
I understand what you are implying but if this stays exactly as it's written one could never submit a TTL value in the form {"_ttl": 1000}. In this case the value is a type long but the field class does not align.

**Comment by ronlepper** [comment]
I'm having a similar problem after upgrading my test suite to use the elasticsearch 2.0 jars.  So how would you necessarily set TTLFieldType so that is not a long field type?  Would it need to be in timestamp form?

**Comment by rjernst** [comment]
@rileytj That should not be the case, `_ttl` is a builtin field. The mapping is not dynamically added, so whether the value is passed as a string or a number in json, it should coerce to a number.

@ronlepper Overriding the builtin `_ttl` field is not supported right now. If you need custom ttl behavior, you will need to implement this yourself. But TTL is generally a bad practice: use time-based indices instead.

**Comment by rileytj** [comment]
https://github.com/elastic/elasticsearch/issues/11882

**Comment by Rajind** [comment]
what is the solution for this ?

**Comment by Rajind** [comment]
I'm getting this exception on field ttl. how can i fix this ? I'm on 2.3

```
MapperParsingException[failed to parse]; nested: IllegalStateException[Mixing up field types: class org.elasticsearch.index.mapper.core.LongFieldMapper$LongFieldType != class org.elasticsearch.index.mapper.internal.TTLFieldMapper$TTLFieldType on field _ttl];
at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:154)
at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:309)
at org.elasticsearch.index.shard.IndexShard.prepareCreate(IndexShard.java:529)
at org.elasticsearch.index.shard.IndexShard.prepareCreateOnPrimary(IndexShard.java:506)
at org.elasticsearch.action.index.TransportIndexAction.prepareIndexOperationOnPrimary(TransportIndexAction.java:215)
at org.elasticsearch.action.index.TransportIndexAction.executeIndexRequestOnPrimary(TransportIndexAction.java:224)
at org.elasticsearch.action.bulk.TransportShardBulkAction.shardIndexOperation(TransportShardBulkAction.java:326)
at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:119)
at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:68)
at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.doRun(TransportReplicationAction.java:639)
at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:279)
at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:271)
at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:75)
at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: Mixing up field types: class org.elasticsearch.index.mapper.core.LongFieldMapper$LongFieldType != class org.elasticsearch.index.mapper.internal.TTLFieldMapper$TTLFieldType on field _ttl
    at org.elasticsearch.index.mapper.FieldMapper.updateFieldType(FieldMapper.java:397)
    at org.elasticsearch.index.mapper.FieldMapper.updateFieldType(FieldMapper.java:53)
    at org.elasticsearch.index.mapper.DocumentParser.parseDynamicValue(DocumentParser.java:622)
    at org.elasticsearch.index.mapper.DocumentParser.parseValue(DocumentParser.java:442)
    at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:262)
    at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:122)
    ... 18 more
```
